libs/media-player/src/lib/audio-annotator.tsx

```diff
@@ -0,0 +1,274 @@
+ */
+export enum AudioMIMEType {
+    mp3 = 'audio/mpeg',
+    audioMp4 = 'audio/mp4',
+    audioOgg = 'audio/ogg',
+    // TODO change this to audio/wav. This requires a migration
+    wav = 'audio/x-wav',
+    audioWebm = 'audio/webm',
+}
+
+const isAudioMIMEType = (input: unknown): input is AudioMIMEType =>
+    Object.values(AudioMIMEType).some((value) => value === (input as AudioMIMEType));
+
+const isValidTimeRangeSelection = (timeRangeSelection: TimeRangeSelection, duration: number) => {
+    const { inPointSeconds, outPointSeconds } = timeRangeSelection;
+
+    if (outPointSeconds >= duration) return false;
+
+    return outPointSeconds > inPointSeconds;
+};
+
+export type TimeRangeSelection = {
+    inPointSeconds: number;
+    outPointSeconds: number;
+};
+
+type TimeRangeSelectionBar = 'inPointSelected' | 'fullSelection';
+
+type UXState = {
+    timeRangeSelectionBar: TimeRangeSelectionBar | null;
+};
+
+const isBarVisible = (stateBar: TimeRangeSelectionBar, { timeRangeSelectionBar }: UXState) =>
+    timeRangeSelectionBar === stateBar;
+
+interface AudioAnnotatorProps {
+    audioUrl: string;
+    mimeType?: AudioMIMEType;
+    onTimeRangeSelected: (timeRangeSelected: TimeRangeSelection | null) => void;
+}
+
+const StyledAudioPlayer = styled('audio')`
+    border-radius: 20px;
+`;
+
+const TimeRangeSelectionVisual = styled(Box)({
+    height: '20px',
+    width: '150px',
+    backgroundColor: '#75ecff',
+    borderRadius: '4px',
+    position: 'absolute',
+});
+
+export const AudioAnnotator = ({
+    audioUrl,
+    mimeType,
+    onTimeRangeSelected,
+}: AudioAnnotatorProps) => {
+    const [uxState, setUxState] = useState<UXState>({
+        timeRangeSelectionBar: null,
+    });
+
+    const audioRef = useRef<HTMLVideoElement>(null);
+
+    const [isPlayHeadMoved, setIsPlayHeadMoved] = useState<boolean>(false);
+
+    const [inPointSeconds, setInPointSeconds] = useState<number | null>(null);
+
+    const [outPointSeconds, setOutPointSeconds] = useState<number | null>(null);
+
+    const [errorMessage, setErrorMessage] = useState<string>('');
+
+    const onTimeUpdate = () => {
+        const currentTime = audioRef.current!.currentTime;
+
+        if (currentTime > 0) {
+            setIsPlayHeadMoved(true);
+        }
+    };
+
+    const markInPoint = () => {
+        setErrorMessage('');
+
+        console.log({ inpoint: audioRef.current!.currentTime });
+
+        setInPointSeconds(audioRef.current!.currentTime);
+
+        toggleTimeRangeSelectionBars('inPointSelected');
+    };
+
+    const markOutPoint = () => {
+        setErrorMessage('');
+
+        console.log({ outpoint: audioRef.current!.currentTime });
+
+        const timeRangeSelection = {
+            inPointSeconds: inPointSeconds as number,
+            outPointSeconds: audioRef.current!.currentTime as number,
+        };
+
+        const duration = audioRef.current!.duration;
+
+        if (!isValidTimeRangeSelection(timeRangeSelection, duration)) {
```

Comment:
I think things will be simpler if you treat the time range as the fundamental state, set this on update, then derive the error state.

Reply:
I tried working with the timeRange as the fundamental state, but it was too awkward working with nested null values in the JSX template.